crates/bevy_pbr/src/render_graph/forward_pipeline/forward.frag

```diff
@@ -24,42 +62,244 @@ layout(set = 1, binding = 0) uniform Lights {
-        v_Uv);
-# endif
-
-# ifndef STANDARDMATERIAL_UNLIT
-    vec3 normal = normalize(v_Normal);
+    vec4 output_color = base_color_factor;
+#ifdef STANDARDMATERIAL_BASE_COLOR_TEXTURE
+    output_color *= texture(sampler2D(StandardMaterial_base_color_texture,
+                                      StandardMaterial_base_color_texture_sampler),
+                            v_Uv);
+#endif
+
+#ifndef STANDARDMATERIAL_UNLIT
+    // calculate non-linear roughness from linear perceptualRoughness
+    float roughness = perceptualRoughnessToRoughness(perceptual_roughness);
+
+    vec3 N = normalize(v_Normal);
+    vec3 V = normalize(CameraPos.xyz - w_Position.xyz);
+
+    // Remapping [0,1] reflectance to F0
+    // See https://google.github.io/filament/Filament.html#materialsystem/parameterization/remapping
+    vec3 F0 = 0.16 * reflectance * reflectance * (1.0 - metallic) + output_color.rgb * metallic;
+
+    // Diffuse strength inversely related to metallicity
+    vec3 diffuseColor = output_color.rgb * (1.0 - metallic);
+
     // accumulate color
-    vec3 color = AmbientColor;
-    for (int i=0; i<int(NumLights.x) && i<MAX_LIGHTS; ++i) {
+    vec3 light_accum = vec3(0.0);
+    for (int i = 0; i < int(NumLights.x) && i < MAX_LIGHTS; ++i) {
         Light light = SceneLights[i];
-        // compute Lambertian diffuse term
-        vec3 light_dir = normalize(light.pos.xyz - v_Position);
-        float diffuse = max(0.0, dot(normal, light_dir));
-        // add light contribution
-        color += diffuse * light.color.xyz;
+
+        vec3 lightDir = light.pos.xyz - w_Position.xyz;
+        vec3 L = normalize(lightDir);
+
+        float rangeAttenuation =
+            getDistanceAttenuation(lightDir, light.inverseRadiusSquared);
+
+        vec3 H = normalize(L + V);
+
+        // Neubelt and Pettineo 2013, "Crafting a Next-gen Material Pipeline for The Order: 1886"
+        float NdotV = max(dot(N, V), 1e-4);
+
+        float NoL = saturate(dot(N, L));
+        float NoH = saturate(dot(N, H));
+        float LoH = saturate(dot(L, H));
+
+        vec3 specular = specular(F0, roughness, H, NdotV, NoL, NoH, LoH);
+        vec3 diffuse = diffuseColor * Fd_Burley(roughness, NdotV, NoL, LoH);
+
+        // Lout = f(v,l) Φ / { 4 π d^2 }⟨n⋅l⟩
+        // where
+        // f(v,l) = (f_d(v,l) + f_r(v,l)) * light_color
+        // Φ is light intensity
+
+        // our rangeAttentuation = 1 / d^2 multiplied with an attenuation factor for smoothing at the edge of the non-physical maximum light radius
+
+        // See https://google.github.io/filament/Filament.html#mjx-eqn-pointLightLuminanceEquation
+        // TODO compensate for energy loss https://google.github.io/filament/Filament.html#materialsystem/improvingthebrdfs/energylossinspecularreflectance
+        light_accum +=
+            ((diffuse + specular) * light.color.rgb) * (light.intensity * 1 / (4 * PI) * rangeAttenuation * NoL);
```

Comment:
The `* 1 / (4 * PI)` is only supposed part of the specular, not the whole lighting equation.

Reply:
And, specifically, the PI is covered by GGX, so it shouldn't be part of it anyway.

Reply:
If I understand this correctly, `f(v,l)` is the BSDF, so both specular and diffuse. But the derivation in the design document doesn't go on to show how the 1/4PI is canceled out, but it doesn't occur in the shaders.

Reply:
I've removed it and added a comment.

Reply:
My understanding has always been that the `1 / PI` is always part of the diffuse and D terms, and the `1 / 4` is part of the Cook-Torrance specular BRDF. Hence why they wouldn't be there.